logins-api/android/library/src/main/java/org/mozilla/sync15/logins/SyncResult.kt

```diff
@@ -123,11 +123,17 @@ class SyncResult<T> {
 
             override fun onException(exception: Exception) {
                 if (exceptionListener == null) {
-                    // Do not swallow thrown exceptions if a listener is not present
-                    throw exception
+                    // Let later handlers in the chain handle the exception.
+                    result.completeFrom(fromException(exception))
+                } else {
+                    try {
+                        result.completeFrom(exceptionListener.onException(exception))
+                    } catch (ex: Exception) {
+                        // Allow users to `throw err` instead of returning `SyncResult.fromException`
```

Comment:
Are we confident that all of our uses "play along"?  Presumably your test suite is exercising all these consumers, or at least the happy paths of these consumers.

Reply:
Yeah, we are. We were doing the wrong thing in a few places, since I assumed this worked like Promise. I'll actually check that the rest of the API behaves this way though -- it probably doesn't now that I think about it...

It would be nice to be able to find out about unhandled exceptions still, but it's more work, I have to think through how to do it and maybe look at what machinery JS vms use (presumably we can emulate it using finalizers somehow)